### PR TITLE
Schematics improvements

### PR DESCRIPTION
**Pull request by skydever**
- [x] Reducer Function needs to be typed
- [x] Allow creating a duck without passing `--project`, should work using `ng generate`
- [x] Document all available parameters for the generation command in the `README.md`

see #5

### CODE REVIEW

**Comment by skydever** [comment]
I am investigating the `--project` parameter and some strange things happened when I created a new app, the schematic did not work out of the box:

* I created a new app `ng new sample-ng-app`, `cd sample-ng-app`
* I added the ducks `npm install @co-it/ngrx-ducks`
* I ran `ng g @co-it/ngrx-ducks:duck my-feature/my-duck` and the following error showed up:
```bash
Schematic input does not validate against the Schema: {"barrel":true,"flat":false,"spec":true}
Errors:

  Data path "" should have required property 'name'.
```

I could remember that I had this error too when I created some schematics for another project. I removed the `"required": ["name"]` option from `schema.json` back then to overcome this issue, and my fallback was the `"x-prompt": "What name would you like to use for the Duck?"` option that is set at the `name` parameter, but this does not work in this case (I removed the `required` property from `schema.json` inside `node_modules`, here `sample-ng-app\node_modules\@co-it\ngrx-ducks-schematics\src\collection\duck`) but then I get the following when I do not pass a `name`:

```bash
ng g @co-it/ngrx-ducks:duck
Cannot read property 'match' of undefined
```

When I pass a `name` it works as expected:
```bash
ng g @co-it/ngrx-ducks:duck my-feature/my-duck
CREATE src/app/my-feature/my-duck/index.ts (65 bytes)
CREATE src/app/my-feature/my-duck/my-duck.duck.spec.ts (333 bytes)
CREATE src/app/my-feature/my-duck/my-duck.duck.ts (441 bytes)
CREATE src/app/my-feature/my-duck/my-duck.state.ts (49 bytes)
```

It seems like the missing `--projects` parameter is not the problem for that case, then I tried that too.

* I added a second app `ng g application second-ng-app`
* Same error if I run `ng g @co-it/ngrx-ducks:duck` (`Cannot read property 'match' of undefined`)
* When I run `ng g @co-it/ngrx-ducks:duck my-awesome-feature/my-duck` it works:
```bash
ng g @co-it/ngrx-ducks:duck my-awesome-feature/my-duck
CREATE src/app/my-awesome-feature/my-duck/index.ts (65 bytes)
CREATE src/app/my-awesome-feature/my-duck/my-duck.duck.spec.ts (333 bytes)
CREATE src/app/my-awesome-feature/my-duck/my-duck.duck.ts (441 bytes)
CREATE src/app/my-awesome-feature/my-duck/my-duck.state.ts (49 bytes)
```
* Passing `--project=second-ng-app` also works:
```bash
ng g @co-it/ngrx-ducks:duck my-hot-feature/my-duck --project=second-ng-app
CREATE projects/second-ng-app/src/app/my-hot-feature/my-duck/index.ts (65 bytes)
CREATE projects/second-ng-app/src/app/my-hot-feature/my-duck/my-duck.duck.spec.ts (333 bytes)
CREATE projects/second-ng-app/src/app/my-hot-feature/my-duck/my-duck.duck.ts (441 bytes)
CREATE projects/second-ng-app/src/app/my-hot-feature/my-duck/my-duck.state.ts (49 bytes)
```

I think because of [this](https://github.com/angular/angular-cli/issues/13010#issuecomment-440302068). I am not sure, but I think the `ng generate` command is handling the default values, like this one that is set for the `project` parameter:
```json
"$default": {
  "$source": "projectName"
}
```
We are depending on `ng generate` to get the correct default value for `project`, but the one from the name seems to be handled outside of `ng generate` (it works using the `schematics` command), this one:
```json
"$default": {
  "$source": "argv",
    "index": 0
}
```
but the `"x-prompt": "What name would you like to use for the Duck?"` is not working for it. At the schematics I made for another project the prompt does show up, but I also set my custom collection as default collection at `angular.json`. I tried that for the ducks schematics too:
```json
  "cli": {
    "defaultCollection": "@co-it/ngrx-ducks-schematics/src/collection.json"
  }
```
but without success when I do not pass a name, same error. With a name I get a result:
```json
ng g duck my-hot-feature/my-duck
CREATE src/app/my-hot-feature/my-duck/index.ts (65 bytes)
CREATE src/app/my-hot-feature/my-duck/my-duck.duck.spec.ts (333 bytes)
CREATE src/app/my-hot-feature/my-duck/my-duck.duck.ts (441 bytes)
CREATE src/app/my-hot-feature/my-duck/my-duck.state.ts (49 bytes)
```

I am not sure what to do at the moment, but I will investigate it further ...

**Comment by skydever** [comment]
something more to notice, seems like my `@angular/cli` is outdated (not my main dev machine):

```bash
ng --version

     _                      _                 ____ _     ___
    / \   _ __   __ _ _   _| | __ _ _ __     / ___| |   |_ _|
   / △ \ | '_ \ / _` | | | | |/ _` | '__|   | |   | |    | |
  / ___ \| | | | (_| | |_| | | (_| | |      | |___| |___ | |
 /_/   \_\_| |_|\__, |\__,_|_|\__,_|_|       \____|_____|___|
                |___/


Angular CLI: 6.2.9
Node: 8.12.0
OS: win32 x64
Angular: 6.1.10
... animations, common, compiler, compiler-cli, core, forms
... http, language-service, platform-browser
... platform-browser-dynamic, router

Package                           Version
-----------------------------------------------------------
@angular-devkit/architect         0.8.9
@angular-devkit/build-angular     0.8.9
@angular-devkit/build-optimizer   0.8.9
@angular-devkit/build-webpack     0.8.9
@angular-devkit/core              0.8.9
@angular-devkit/schematics        0.8.9
@angular/cli                      6.2.9
@ngtools/webpack                  6.2.9
@schematics/angular               0.8.9
@schematics/update                0.8.9
rxjs                              6.2.2
typescript                        2.9.2
webpack                           4.16.4
```

**Comment by skydever** [comment]
no support for prompts in `6.2.9`, I will try again with the latest version ...

**Comment by skydever** [comment]
* [x] `ng g @co-it/ngrx-ducks:duck my-feature/my-duck`
* [x] `ng g @co-it/ngrx-ducks:duck` - the prompt shows up
* add a second app
* [x] `ng g @co-it/ngrx-ducks:duck`
* [x] `ng g @co-it/ngrx-ducks:duck --project=second-app`
* [x] `ng g @co-it/ngrx-ducks:duck feature-one/super-duck --project=second-app`
* [x] `ng g @co-it/ngrx-ducks:duck feature-two/super-duck`

... looks good concering `required`, `project` and the prompt. Did you have issues with the `project` parameter, did you create the angular workspace (multiple projects) in another way? I will try with a **@nrwl/nx** workspace tomorrow.

**Comment by skydever** [comment]
```bash
Angular CLI: 7.3.4
Node: 8.12.0
OS: win32 x64
Angular: 7.2.7
... animations, common, compiler, compiler-cli, core, forms
... language-service, platform-browser, platform-browser-dynamic
... router

Package                           Version
-----------------------------------------------------------
@angular-devkit/architect         0.13.4
@angular-devkit/build-angular     0.13.4
@angular-devkit/build-optimizer   0.13.4
@angular-devkit/build-webpack     0.13.4
@angular-devkit/core              7.3.4
@angular-devkit/schematics        7.3.4
@angular/cli                      7.3.4
@ngtools/webpack                  7.3.4
@schematics/angular               7.3.4
@schematics/update                0.13.4
rxjs                              6.3.3
typescript                        3.2.4
webpack                           4.29.0
```

**Comment by skydever** [comment]
testing this commands would be great, I think I can add tests to check if the files based on the properties exist at the right place after execution ... not sure if that would catch incompatibility with `ng generate` as well?

**Comment by GregOnNet** [comment]
Thanks for the detailed report for this issue.
Can you summarize the problems that still exist and which you were able to solve on your own?

**Comment by skydever** [comment]
My globally installed Angular CLI was outdated (`6.2.9`) when I started to test the schematics. The version did not have support for prompts, this feature came with `7` - that was the reason for the problems I was facing. After upgrading to the lastest Angular CLI version (`7.3.4`), I had no problems at all. Today I tried the commands within a **@nrwl/nx workspace** too and everything worked as expected.

### `--project` parameter

* It is not required to explicitly set a `project` when the schematics are executed via `ng generate`, it will pick the `defaultProject` from `angular.json` -> *this already works with @co-it/ngrx-ducks 7.6.0*
  * Passing a `project` is only required if the schematic is executed via the `schematics` command, as described [here](https://github.com/angular/angular-cli/issues/13010#issuecomment-440302068) - or if no `defaultProject` is set at `angular.json`.

*Can you confirm this behavior?*

### `name` parameter

* If no name is passed to the execution (eg: `ng g @co-it/ngrx-ducks:duck`) a prompt will show up asking for the name of the duck: `What name would you like to use for the Duck?` -> *this also already works with @co-it/ngrx-ducks 7.6.0*

### Typed reducer function

I added the types to the reducer function.

### Documentation

I added the possible parameters and a description to the `README.md`.


This should cover all the tasks from #5, or dir I miss something?

**Comment by GregOnNet** [comment]
> Passing a project is only required if the schematic is executed via the schematics command, as described here - or if no defaultProject is set at angular.json.

Aah, I was missing this bit about the `schematics` cli.
Yes, it works with the real CLI. :-)

---

I will double check everything this evening.
Afterwards, I will do a patch release. 👍

**Comment by GregOnNet** [comment]
@skydever Once again, thanks a lot for your input!
I added your changes and released version `7.6.1`.

**Comment by skydever** [comment]
great, I am happy that it is working 😃
you are very welcome 👍
